### PR TITLE
Enable pinvoke and resource usage analyzer

### PR DESCRIPTION
**Pull request by pallavit**

### CODE REVIEW

**Comment by pallavit** [comment]
/cc: @weshaggard, @tarekgh Could you please take a look at this.

File: src/System.Security.Cryptography.Csp/src/PinvokeAnalyzerExceptionList.analyzerdata [hunk 1]
```diff
@@ -0,0 +1,19 @@
+<!-- This is a bug. This dll supports .Net Standard 1.5 and so should be using APISet dependencies -->
+Advapi32.dll!CryptGetDefaultProviderW
```
Comment (pallavit) [inline src/System.Security.Cryptography.Csp/src/PinvokeAnalyzerExceptionList.analyzerdata]: Creating a tracking bug for this.

File: src/System.Security.Cryptography.X509Certificates/src/PinvokeAnalyzerExceptionList.analyzerdata [hunk 2]
```diff
@@ -0,0 +1,5 @@
+<!-- This is a bug. This dll supports .NetStandard 1.5 and so should be using APiSet dependency -->
+advapi32.dll!CryptAcquireContextW
+
+<!--Devdiv:1201818 BCrypt* and friends are exposed via different dlls in store vs OneCore profiles. -->
+bcrypt.dll!BCryptDestroyKey
```
Comment (pallavit) [inline src/System.Security.Cryptography.X509Certificates/src/PinvokeAnalyzerExceptionList.analyzerdata]: Creating a tracking bug for this as well

**Comment by tarekgh** [comment]
LGTM

File: src/System.Data.SqlClient/src/PinvokeAnalyzerExceptionList.analyzerdata [hunk 3]
```diff
@@ -0,0 +1,27 @@
+<!-- Not a bug. Sni.dll is a native component that ship as part of the package -->
```
Comment (stephentoub) [inline src/System.Data.SqlClient/src/PinvokeAnalyzerExceptionList.analyzerdata]: Nit: "that ships"

File: src/System.IO.Compression/src/PinvokeAnalyzerExceptionList.analyzerdata [hunk 4]
```diff
@@ -0,0 +1,8 @@
+<!-- Not a bug. clrcompression.dll is a native component that ship as part of the package -->
```
Comment (stephentoub) [inline src/System.IO.Compression/src/PinvokeAnalyzerExceptionList.analyzerdata]: Nit: "that ships"

File: src/System.Reflection.Metadata/src/PinvokeAnalyzerExceptionList.analyzerdata [hunk 5]
```diff
@@ -0,0 +1,3 @@
+<!-- This Pinvoke is allowed on windows, however the dll is built as AnyCPU -->
```
Comment (stephentoub) [inline src/System.Reflection.Metadata/src/PinvokeAnalyzerExceptionList.analyzerdata]: Nit: windows => Windows, Pinvoke =< P/Invoke, dll => DLL

File: src/System.Security.Cryptography.Csp/src/PinvokeAnalyzerExceptionList.analyzerdata [hunk 6]
```diff
@@ -0,0 +1,19 @@
+<!-- This is a bug. This dll supports .Net Standard 1.5 and so should be using APISet dependencies -->
```
Comment (stephentoub) [inline src/System.Security.Cryptography.Csp/src/PinvokeAnalyzerExceptionList.analyzerdata]: Nit: .Net => .NET

**Comment by stephentoub** [comment]
Some nits, otherwise LGTM

**Comment by weshaggard** [comment]
LGTM

File: src/System.Reflection.Metadata/src/System/Reflection/Internal/Utilities/FileStreamReadLightUp.cs [hunk 7]
```diff
@@ -156,7 +156,7 @@ IntPtr overlapped
             );
 
             // older Windows systems:
-            [DllImport(@"kernel32.dll", EntryPoint = "ReadFile", ExactSpelling = true, SetLastError = true)]
+            [DllImport(@"api-ms-win-core-file-l1-1-0.dll", EntryPoint = "ReadFile", ExactSpelling = true, SetLastError = true)]
```
Comment (jkotas) [inline src/System.Reflection.Metadata/src/System/Reflection/Internal/Utilities/FileStreamReadLightUp.cs]: cc @tmat @nguerrera 

Does this change break System.Reflection.Metadata when running on Win7 against desktop?

Comment (tmat) [inline src/System.Reflection.Metadata/src/System/Reflection/Internal/Utilities/FileStreamReadLightUp.cs]: +1, why is this changed?

Comment (stephentoub) [inline src/System.Reflection.Metadata/src/System/Reflection/Internal/Utilities/FileStreamReadLightUp.cs]: I believe @pallavit changed it because it was getting flagged by the API validation analyzer.  But that appears to be by design, so a follow-up change should probably revert this line and add an exemption, since it shouldn't end up getting called if ReadFileModern can be used successfully.

Comment (weshaggard) [inline src/System.Reflection.Metadata/src/System/Reflection/Internal/Utilities/FileStreamReadLightUp.cs]: Just the presence of it will block any UWP application using this library from passing WACK and be rejected from store submission. So we need to evaluate whether or not we want to support this library on UWP and if we do then something else will need to be done.

Comment (jkotas) [inline src/System.Reflection.Metadata/src/System/Reflection/Internal/Utilities/FileStreamReadLightUp.cs]: There is no point in supporting System.Reflection.Metadata for UWP.

Comment (nguerrera) [inline src/System.Reflection.Metadata/src/System/Reflection/Internal/Utilities/FileStreamReadLightUp.cs]: FWIW, at least one user was using it in a windows 8 store app (no idea why), but a bug was reported via nugget feedback and subsequently fixed. I recall that ReadFile via kernel32.dll was actually allowed on win8 when I checked with @ericstj. Did that change with UWP?

Comment (nguerrera) [inline src/System.Reflection.Metadata/src/System/Reflection/Internal/Utilities/FileStreamReadLightUp.cs]: I also recall Roslyn folks pushing for crypto in UWP implying they eventually wanted Roslyn on UWP, which would require S.R.M on UWP. cc @jaredpar

Comment (nguerrera) [inline src/System.Reflection.Metadata/src/System/Reflection/Internal/Utilities/FileStreamReadLightUp.cs]: Can we revert this for now and track if/how to address in a separate bug?

Comment (nydehi) [inline src/System.Reflection.Metadata/src/System/Reflection/Internal/Utilities/FileStreamReadLightUp.cs]: I can tell you one reason.   A client wanted to use bluetooth as a
parameter boundary to turn off networking for surface tablets.
In UWP you can detect bluetooth but not turn off networking. but in regular
windows on tablet win8/10 you can not detect bluetooth.(seriously try it)
 so a file somewhere would be a sentinel to turn on off networking.

Sincerely,
J. John Mele
631 615 7213

linkedin profile: http://www.linkedin.com/in/joemele
Youseful Software
www.youseful.com

Comment (pallavit) [inline src/System.Reflection.Metadata/src/System/Reflection/Internal/Utilities/FileStreamReadLightUp.cs]: > Can we revert this for now and track if/how to address in a separate bug?

Note that both Pinvoke's are incorrect i.e. kernel32!ReadFile and api-ms-win-core-file-l1-1-0!ReadFile since this library is built as AnyCPU while these methods are windows specific. I added exception for api-ms-win-core-file-l1-1-0!ReadFile, however unfortunately,  as the infrastructure stands as of now, I can't add exception to both entries. We currently only support one methodName exception (there is really no reason, other than in this special case, where we want to support multiple module name's for the same entry point) and honestly, I don't think we should add that support in the infrastructure either.

The right fix would be to separate the Desktop implementation and make this library windows specific (or remove windows Pinvokes from the x-plat implementation). A temporary workaround would be to disable the CodeAnalyzers for this project (however note that all code analyzers will be disabled this way). May be adding a separate rule set or property will work but I have not tried that yet.

Comment (tmat) [inline src/System.Reflection.Metadata/src/System/Reflection/Internal/Utilities/FileStreamReadLightUp.cs]: This API should really be provided by System.IO.FileSystem.

File: src/System.Security.Cryptography.Csp/src/PinvokeAnalyzerExceptionList.analyzerdata (hunk 1 above)
Comment (pallavit) [inline src/System.Security.Cryptography.Csp/src/PinvokeAnalyzerExceptionList.analyzerdata]: Issue #7550 created.

File: src/System.Reflection.Metadata/src/System/Reflection/Internal/Utilities/FileStreamReadLightUp.cs (hunk 7 above)
Comment (tmat) [inline src/System.Reflection.Metadata/src/System/Reflection/Internal/Utilities/FileStreamReadLightUp.cs]: @pallavit Please undo the change to this file and file a bug to follow up. If this is flagged by a Roslyn analyzer then a suppression pragma can be added for this file.

Comment (pallavit) [inline src/System.Reflection.Metadata/src/System/Reflection/Internal/Utilities/FileStreamReadLightUp.cs]: >  If this is flagged by a Roslyn analyzer then a suppression pragma can be added for this file.

I forgot this option. Will do. Who should I assign the tracking issue?

Comment (nguerrera) [inline src/System.Reflection.Metadata/src/System/Reflection/Internal/Utilities/FileStreamReadLightUp.cs]: > Note that both Pinvoke's are incorrect

The P/Invokes are guarded by light-up. If both fail to work we fall back to a slower code path. So while I'm not saying that there isn't a genuine WACK issue, I would like to clarify that there is no x-plat (meaning non-Windows) bug here.

>The right fix would be to separate the Desktop implementation and make this library windows specific (or remove windows Pinvokes from the x-plat implementation)

I would really like to avoid this if at all possible. A lib like S.R.M shouldn't need to build separately for Windows and Unix when it is so far removed from the OS in its function.

> This API should really be provided by System.IO.FileSystem.

Yes. I filed #4786 a while ago. Unfortunately, no addition to the BCL can handle the optimization if we want it to continue to kick in on downlevel .NET Framework and Windows.

Comment (pallavit) [inline src/System.Reflection.Metadata/src/System/Reflection/Internal/Utilities/FileStreamReadLightUp.cs]: > would really like to avoid this if at all possible. A lib like S.R.M shouldn't need to build separately for Windows and Unix when it is so far removed from the OS in its function.

The implementation currently has OS specific calls and so for now it can't be AnyCPU really. However depending on how #4786 is fixed it might change :). I will leave the right fix to you guys :)

Comment (nguerrera) [inline src/System.Reflection.Metadata/src/System/Reflection/Internal/Utilities/FileStreamReadLightUp.cs]: > The implementation currently has OS specific calls and so for now it can't be AnyCPU really. However depending on how #4786 is fixed it might change :). I will leave the right fix to you guys :)

Again, we handle the case where the P/Invoke entry point is not found. Many libs in the ecosystem do this too. The issue isn't a problem for non-Windows platforms. Satisfying WACK is the only problem here.

Comment (nguerrera) [inline src/System.Reflection.Metadata/src/System/Reflection/Internal/Utilities/FileStreamReadLightUp.cs]: Also, when you say "any CPU" here it really means "any OS" in this context, right? I hope nothing has changed to require architecture-specific managed libs in order to interop with the OS.

Comment (pallavit) [inline src/System.Reflection.Metadata/src/System/Reflection/Internal/Utilities/FileStreamReadLightUp.cs]: I do mean AnyOS. My bad.
